storops/unity/calculator.py

```diff
@@ -295,6 +313,254 @@ def _sp_pct(path, curr, prev):
+        ret = delta / curr.diff_seconds(prev)
+    else:
+        ret = IdValues()
+    return ret
+
+
+@metric_calculator
+def system_total_delta_ps(path, prev, curr):
+    if len(path) != 2:
+        raise ValueError('takes in "reads" and "writes" counter.')
+
+    part1, part2 = path
+    return system_delta_ps(part1, prev, curr) + system_delta_ps(part2, prev,
+                                                                curr)
+
+
+@metric_calculator
+def disk_response_time(path, prev, curr):
+    if len(path) != 5:
+        raise ValueError(
+            'takes in "busyTicks", "sumArrivalQueueLength", "reads", '
+            '"writes" and "coreCount" counter.')
+
+    part1, part2, part3, part4, part5 = path
+    if all_not_none(prev, curr):
+        prev_part1 = prev.by_path(part1)
+        prev_part2 = prev.by_path(part2)
+        prev_part3 = prev.by_path(part3)
+        prev_part4 = prev.by_path(part4)
+        prev_part5 = prev.by_path(part5)
+
+        curr_part1 = curr.by_path(part1)
+        curr_part2 = curr.by_path(part2)
+        curr_part3 = curr.by_path(part3)
+        curr_part4 = curr.by_path(part4)
+        curr_part5 = curr.by_path(part5)
+
+        delta_part1 = curr_part1.sum_sp() - prev_part1.sum_sp()
+        delta_part2 = curr_part2.sum_sp() - prev_part2.sum_sp()
+        delta_part3 = curr_part3.sum_sp() - prev_part3.sum_sp()
+        delta_part4 = curr_part4.sum_sp() - prev_part4.sum_sp()
+
+        delta_other1 = curr_part3.combine_numeric_values(
+            curr_part5) - prev_part3.combine_numeric_values(prev_part5)
+        delta_other2 = curr_part4.combine_numeric_values(
+            curr_part5) - prev_part4.combine_numeric_values(prev_part5)
+
+        ret = delta_part1 * delta_part2 / ((delta_part3 + delta_part4) *
+                                           (delta_other1 + delta_other2))
+
+    else:
+        ret = IdValues()
+
+    return ret
+
+
+@metric_calculator
+def disk_queue_length(path, prev, curr):
+    if len(path) != 3:
+        raise ValueError(
+            'takes in "sumArrivalQueueLength", "reads" and '
+            '"writes" counter.')
+
+    part1, part2, part3 = path
+    if all_not_none(prev, curr):
+        prev_part1 = prev.by_path(part1)
+        prev_part2 = prev.by_path(part2)
+        prev_part3 = prev.by_path(part3)
+
+        curr_part1 = curr.by_path(part1)
+        curr_part2 = curr.by_path(part2)
+        curr_part3 = curr.by_path(part3)
+
+        delta_part1 = curr_part1.sum_sp() - prev_part1.sum_sp()
+        delta_part2 = curr_part2.sum_sp() - prev_part2.sum_sp()
+        delta_part3 = curr_part3.sum_sp() - prev_part3.sum_sp()
+
+        ret = delta_part1 / (delta_part2 + delta_part3)
+
+    else:
+        ret = IdValues()
+
+    return ret
+
+
+@metric_calculator
+def lun_response_time(path, prev, curr):
+    if len(path) != 3:
+        raise ValueError(
+            'takes in "totalIoTime", "reads" and "writes" counter.')
+
+    part1, part2, part3 = path
```

Comment:
This algorithm looks the same as line 392 ~ 409.  If they are the same, you could extract the common logic to a function to avoid duplication.

Reply:
Done, write a new function to extract the common logic.